fvm/src/call_manager/default.rs

```diff
@@ -412,6 +405,62 @@ where
         Ok(())
     }
 
+    fn upgrade_actor<K: Kernel<CallManager = Self>>(
+        &mut self,
+        actor_id: ActorID,
+        new_code_cid: Cid,
+        params: Option<Block>,
+    ) -> Result<InvocationResult> {
+        let origin = self.origin;
+
+        let state = self
+            .state_tree_mut()
+            .get_actor(actor_id)?
+            .ok_or_else(|| syscall_error!(NotFound; "actor not found: {}", actor_id))?;
+
+        // store the code cid of the calling actor before running the upgrade entrypoint
+        // in case it was changed (which could happen if the target upgrade entrypoint
+        // sent a message to this actor which in turn called upgrade)
+        let code = state.code;
+
+        // update the code cid of the actor to new_code_cid
+        self.state_tree_mut().set_actor(
+            origin,
+            ActorState::new(
+                new_code_cid,
+                state.state,
+                state.balance,
+                state.sequence,
+                None,
+            ),
+        );
+
+        // run the upgrade entrypoint
+        let result = self.send::<K>(
+            actor_id,
+            Address::new_id(actor_id),
+            Entrypoint::Upgrade(UpgradeInfo { old_code_cid: code }),
```

Comment:
Exposing _both_ an `upgrade_actor` method _and_ a send method that can take an `EntryPoint::Upgrade` is a bit strange. I'd rename `send` to `call_actor` and either:

1. Make `call_actor` public, moving the "upgrade" logic into the kernel (if reasonable, I haven't tried it?).
2. Make `call_actor` private and export a separate `send` method. In this case, I'd make `with_transaction` private again as we'd be handling the transaction logic internally.

I'm not entirely sure which way to go. IMO, the second option is "cleaner" from an API perspective (hides internals). However, the first one is more flexible and moves more logic into the kernel (which, IMO, is a good thing). I.e., it means that the call manager doesn't actually need to know about "upgrades".

Reply:
If we go with option 1, I'd make `Entrypoint` a _trait_ so we can define new entrypoints without touching the call manager.